source/grid/grid_generator.cc

```diff
@@ -749,6 +749,63 @@ namespace GridGenerator
     tria.set_all_manifold_ids_on_boundary(0);
   }
 
+
+
+  template <int dim>
+  void
+  hyper_quadrilateral (Triangulation<dim> &tria,
+                       const std::vector<Point<dim> > &vertices,
+                       const bool colorize)
+  {
+    Assert (vertices.size() == std::pow(2., dim),
+            ExcMessage("Wrong number of vertices."));
+
+    // Check that the order of the vertices makes sense
+    Assert (vertices[1][0]>=vertices[0][0],
+            ExcMessage("vertices[1][0] must be greater or equal than vertices[1][0]"));
+#if dim > 1
+    Assert (vertices[3][0]>=vertices[2][0],
+            ExcMessage("vertices[3][0] must be greater or equal than vertices[2][0]"));
+    Assert (vertices[2][1]>=vertices[0][1],
+            ExcMessage("vertices[2][1] must be greater or equal than vertices[0][1]"));
+    Assert (vertices[3][1]>=vertices[1][1],
+            ExcMessage("vertices[3][1] must be greater or equal than vertices[1][1]"));
+#endif
+#if dim == 3
+    Assert (vertices[5][0]>=vertices[4][0],
+            ExcMessage("vertices[5][0] must be greater or equal than vertices[4][0]"));
+    Assert (vertices[7][0]>=vertices[6][0],
+            ExcMessage("vertices[7][0] must be greater or equal than vertices[6][0]"));
+    Assert (vertices[6][1]>=vertices[4][1],
+            ExcMessage("vertices[6][1] must be greater or equal than vertices[4][1]"));
+    Assert (vertices[7][1]>=vertices[5][1],
+            ExcMessage("vertices[7][1] must be greater or equal than vertices[5][1]"));
+    Assert (vertices[4][2]>=vertices[0][2],
+            ExcMessage("vertices[4][2] must be greater or equal than vertices[0][2]"));
+    Assert (vertices[5][2]>=vertices[1][2],
+            ExcMessage("vertices[5][2] must be greater or equal than vertices[1][2]"));
+    Assert (vertices[6][2]>=vertices[2][2],
+            ExcMessage("vertices[6][2] must be greater or equal than vertices[2][2]"));
+    Assert (vertices[7][2]>=vertices[3][2],
+            ExcMessage("vertices[7][2] must be greater or equal than vertices[3][2]"));
```

Comment:
Why do these assertions make sense? Imagine you wanted to just create a rotated square, then several of these assertions are going to trigger.

You make an assumption here about the alignment of the cell with some kind of coordinate system. I think you should drop this and instead just test that the volume of the cell is positive. I think there are functions in `GridTools` that check that.